src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        currentState = Object.keys(currentState).reduce((newState, key) => {
+          if (!action.keysToRemove.includes(key)) {
+            newState[key] = currentState[key];
+          }
+
+          return newState;
+        }, {});
+        break;
+
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
```

Comment:
The switch statement is used correctly here to handle different action types and there is a default case for error handling, which is good for robustness. However, the action of cloning the currentState into stateHistory is repeated in every case. This can be done once after the switch statement to adhere to the DRY principle.